File: web/src/service/notification.ts
import notificationApi from '@/api/notification'
import { NormalResponse, DeleteNotificationRecordRequest, InifiniteScrollPagnitionNotificationRecord, NotificationRecordDetailRequest, ReadNotificationRecordRequest, SearchNotificationRecordRequest, AddNotificationSourceRequest, UpdateNotificationSourceRequest, AddNotificationTargetRequest, UpdateNotificationTargetRequest, AddNotificationTaskRequest, DeleteNotificationTaskRequest, UpdateNotificationTaskRequest, NotificationTask, NotificationTaskDetailRequest, PaginationNotificationTask, PageableRequest, GetNotificationSourceRelatedTaskResponse, GetNotificationTargetRelatedTaskResponse, GetNotificationTargetRelatedTaskRequest, GetNotificationSourceRelatedTaskRequest, SearchNotificationSourceRequest, InifiniteScrollPagnitionNotificationSource, NotificationSourceForkRequest, NotificationTargetForkRequest, DeleteNotificationSourceRequest, NotificationSourceDetailRequest, NotificationSourceDetail, DeleteNotificationTargetRequest, NotificationTargetDetailRequest, SearchNotificationTargetRequest, InifiniteScrollPagnitionNotificationTarget, NotificationTargetsProvidedResponse, NotificationSourcesProvidedResponse, NotificationTargetDetail, NotificationSourcesUsableResponse, NotificationTargetsUsableResponse, EmailTargetSendCodeRequest, UserPublicInfo } from '@/generated';
import { request } from '@/lib/request';

export type NotificationTemplateParameterBinding = {
	source_type: 'event' | 'static';
	attribute_key?: string | null;
	static_value?: string | null;
};

export type NotificationTemplateParameter = {
	id: number;
	key: string;
	label: string;
	description?: string | null;
	value_type: string;
	required: boolean;
	default_value?: string | null;
};

export type NotificationTemplateItem = {
	id: number;
	uuid: string;
	creator_id: number;
	is_public: boolean;
	name: string;
	description?: string | null;
	title_template?: string | null;
	content_template?: string | null;
	link_template?: string | null;
	cover_template?: string | null;
	create_time: string;
	update_time?: string | null;
	creator: UserPublicInfo;
	is_forked?: boolean | null;
	parameters: NotificationTemplateParameter[];
};

export type TriggerEventAttribute = {
	id: number;
	key: string;
	label: string;
	label_zh?: string | null;
	description?: string | null;
	description_zh?: string | null;
	value_type: string;
	required: boolean;
};

export type TriggerEventItem = {
	id: number;
	uuid: string;
	name: string;
	name_zh: string;
	description?: string | null;
	description_zh?: string | null;
	create_time: string;
	update_time?: string | null;
	attributes: TriggerEventAttribute[];
};

export type NotificationTemplatesListResponse = {
	data: NotificationTemplateItem[];
};

export type TriggerEventsListResponse = {
	data: TriggerEventItem[];
};

export type NotificationTemplateUpsertRequest = {
	notification_template_id?: number;
	name: string;
	is_public: boolean;
	description?: string;
	title_template: string;
	content_template?: string;
	link_template?: string;
	cover_template?: string;
	parameters: Array<{
		key: string;
		label: string;
		description?: string;
		value_type: string;
		required: boolean;
		default_value?: string;
	}>;
};

export type NotificationTemplateForkRequest = {
	notification_template_id: number;
	status: boolean;
};

export type SearchNotificationTemplateRequest = {
	keyword?: string;
	start?: number;
	limit: number;
};

export type InifiniteScrollPagnitionNotificationTemplate = {
	total: number;
	start?: number | null;
	limit: number;
	has_more: boolean;
	elements: NotificationTemplateItem[];
	next_start?: number | null;
};

export type NotificationTaskPayload = Omit<AddNotificationTaskRequest, 'notification_template_id'> & {
	notification_template_id?: number;
	notification_template_bindings?: Record<string, NotificationTemplateParameterBinding>;
};

export type NotificationTaskUpdatePayload = Omit<UpdateNotificationTaskRequest, 'notification_template_id'> & {
	notification_template_id?: number;
	notification_template_bindings?: Record<string, NotificationTemplateParameterBinding>;
};

export type NotificationTaskDetailData = NotificationTask & {
	notification_template_bindings?: Record<string, NotificationTemplateParameterBinding>;
};

export const forkNotificationSource = async (data: NotificationSourceForkRequest): Promise<NormalResponse> => {
    return await request(notificationApi.forkNotificationSource, {
        data
    })
}

export const sendEmailNotificationTargetCode = async (data: EmailTargetSendCodeRequest): Promise<NormalResponse> => {
    return await request(notificationApi.sendEmailNotificationTargetCode, {
        data
    })
}

export const forkNotificationTarget = async (data: NotificationTargetForkRequest): Promise<NormalResponse> => {
    return await request(notificationApi.forkNotificationTarget, {
        data
    })
}

export const getNotificationSourceRelatedTasks = async (data: GetNotificationSourceRelatedTaskRequest): Promise<GetNotificationSourceRelatedTaskResponse> => {
    return await request(notificationApi.getNotificationSourceRelatedTasks, {
        data
    })
}

export const getNotificationTargetRelatedTasks = async (data: GetNotificationTargetRelatedTaskRequest): Promise<GetNotificationTargetRelatedTaskResponse> => {
    return await request(notificationApi.getNotificationTargetRelatedTasks, {
        data
    })
}

export const getTriggerEvents = async (): Promise<TriggerEventsListResponse> => {
    return await request(notificationApi.getTriggerEvents)
}

export const getProvidedNotificationSources = async (): Promise<NotificationSourcesProvidedResponse> => {
    return await request(notificationApi.getProvidedNotificationSources)
}

export const getProvidedNotificationTargets = async (): Promise<NotificationTargetsProvidedResponse> => {
    return await request(notificationApi.getProvidedNotificationTargets)
}

export const deleteNotificationRecords = async (data: DeleteNotificationRecordRequest): Promise<NormalResponse> => {
    return await request(notificationApi.deleteNotificationRecords, {
        data
    })
}

export const searchNotificationRecords = async (data: SearchNotificationRecordRequest): Promise<InifiniteScrollPagnitionNotificationRecord> => {
    return await request(notificationApi.searchNotificationRecords, {
        data
    })
}

export const readNotificationRecords = async (data: ReadNotificationRecordRequest): Promise<NormalResponse> => {
    return await request(notificationApi.readNotificationRecord, {
        data
    })
}

export const getNotificationRecordDetail = async (data: NotificationRecordDetailRequest): Promise<Notification> => {
    return await request(notificationApi.getNotificationRecordDetail, {
        data
    })
}

export const readAllNotificationRecords = async (): Promise<NormalResponse> => {
    return await request(notificationApi.readAllNotificationRecords)
}

export const addNotificationSource = async (data: AddNotificationSourceRequest): Promise<NormalResponse> => {
    return await request(notificationApi.addNotificationSource, {
        data
    })
}

export const deleteNotificationSource = async (data: DeleteNotificationSourceRequest): Promise<NormalResponse> => {
    return await request(notificationApi.deleteNotificationSource, {
        data
    })
}

export const getMineNotificationSourceDetail = async (data: NotificationSourceDetailRequest): Promise<NotificationSourceDetail> => {
    return await request(notificationApi.getNotificationSourceDetail, {
        data
    })
}

export const getMineNotificationSources = async (data: SearchNotificationSourceRequest): Promise<InifiniteScrollPagnitionNotificationSource> => {
    return await request(notificationApi.getMineNotificationSources, {
        data
    })
}

export const getUsableNotificationSources = async (): Promise<NotificationSourcesUsableResponse> => {
    return await request(notificationApi.getUsableNotificationSources)
}

export const updateNotificationSource = async (data: UpdateNotificationSourceRequest): Promise<NormalResponse> => {
    return await request(notificationApi.updateNotificationSource, {
        data
    })
}

export const addNotificationTarget = async (data: AddNotificationTargetRequest): Promise<NormalResponse> => {
    return await request(notificationApi.addNotificationTarget, {
        data
    })
}

export const deleteNotificationTarget = async (data: DeleteNotificationTargetRequest): Promise<NormalResponse> => {
    return await request(notificationApi.deleteNotificationTarget, {
        data
    })
}

export const getMineNotificationTargetDetail = async (data: NotificationTargetDetailRequest): Promise<NotificationTargetDetail> => {
    return await request(notificationApi.getNotificationTargetDetail, {
        data
    })
}

export const getMineNotificationTargets = async (data: SearchNotificationTargetRequest): Promise<InifiniteScrollPagnitionNotificationTarget> => {
    return await request(notificationApi.getMineNotificationTargets, {
        data
    })
}

export const getUsableNotificationTargets = async (): Promise<NotificationTargetsUsableResponse> => {
    return await request(notificationApi.getUsableNotificationTargets)
}

export const updateNotificationTarget = async (data: UpdateNotificationTargetRequest): Promise<NormalResponse> => {
    return await request(notificationApi.updateNotificationTarget, {
        data
    })
}

export const addNotificationTask = async (data: NotificationTaskPayload): Promise<NormalResponse> => {
    return await request(notificationApi.addNotificationTask, {
        data
    })
}

export const deleteNotificationTask = async (data: DeleteNotificationTaskRequest): Promise<NormalResponse> => {
    return await request(notificationApi.deleteNotificationTask, {
        data
    })
}

export const updateNotificationTask = async (data: NotificationTaskUpdatePayload): Promise<NormalResponse> => {
    return await request(notificationApi.updateNotificationTask, {
        data
    })
}

export const getMineNotificationTask = async (data: PageableRequest): Promise<PaginationNotificationTask> => {
    return await request(notificationApi.getMineNotificationTask, {
        data
    })
}

export const getNotificationTaskDetail = async (data: NotificationTaskDetailRequest): Promise<NotificationTaskDetailData> => {
    return await request(notificationApi.getNotificationTaskDetail, {
        data
    })
}

export const getNotificationTemplate = async (): Promise<NotificationTemplatesListResponse> => {
    return await request(notificationApi.getNotificationTemplate)
}

export const getNotificationTemplatesCommunity = async (
	data: SearchNotificationTemplateRequest,
): Promise<InifiniteScrollPagnitionNotificationTemplate> => {
	return await request(notificationApi.getNotificationTemplatesCommunity, {
		data,
	});
};

export const upsertNotificationTemplate = async (
	data: NotificationTemplateUpsertRequest,
): Promise<NormalResponse> => {
	return await request(notificationApi.upsertNotificationTemplate, {
		data,
	});
};

export const deleteNotificationTemplate = async (data: {
	notification_template_id: number;
}): Promise<NormalResponse> => {
	return await request(notificationApi.deleteNotificationTemplate, {
		data,
	});
};

export const forkNotificationTemplate = async (
	data: NotificationTemplateForkRequest,
): Promise<NormalResponse> => {
	return await request(notificationApi.forkNotificationTemplate, {
		data,
	});
};

File: api/schemas/notification.py
from datetime import datetime
from typing import Literal

from pydantic import Field

from .base import BaseModel
from .user import UserPublicInfo

PUBLIC_PAGINATION_LIMIT = 20

class IOSTargetChangeCodeStatusRequest(BaseModel):
    code_uuid: str
    status: Literal["scanned", "confirm", "canceled"]
    device_token: str | None = None

class EmailTargetSendCodeRequest(BaseModel):
    email: str | None = None

class NotificationSourceForkRequest(BaseModel):
    notification_source_id: int
    status: bool
    
class NotificationTargetForkRequest(BaseModel):
    notification_target_id: int
    status: bool

class NotificationTemplateForkRequest(BaseModel):
    notification_template_id: int
    status: bool

class SearchNotificationTargetRequest(BaseModel):
    keyword: str | None = None
    start: int | None = None
    limit: int = Field(default=10, le=PUBLIC_PAGINATION_LIMIT)

class SearchNotificationSourceRequest(BaseModel):
    keyword: str | None = None
    start: int | None = None
    limit: int = Field(default=10, le=PUBLIC_PAGINATION_LIMIT)

class SearchNotificationTemplateRequest(BaseModel):
    keyword: str | None = None
    start: int | None = None
    limit: int = Field(default=10, le=PUBLIC_PAGINATION_LIMIT)

class NotificationTaskBaseInfo(BaseModel):
    id: int
    title: str
    create_time: datetime
    update_time: datetime | None

class GetNotificationTargetRelatedTaskRequest(BaseModel):
    notification_target_id: int

class GetNotificationTargetRelatedTaskResponse(BaseModel):
    data: list[NotificationTaskBaseInfo]

class GetNotificationSourceRelatedTaskRequest(BaseModel):
    notification_source_id: int

class GetNotificationSourceRelatedTaskResponse(BaseModel):
    data: list[NotificationTaskBaseInfo]

class TriggerEvent(BaseModel):
    id: int
    uuid: str
    name: str
    name_zh: str
    description: str | None
    description_zh: str | None
    create_time: datetime
    update_time: datetime | None
    attributes: list['TriggerEventAttribute'] = []

class TriggerEventAttribute(BaseModel):
    id: int
    key: str
    label: str
    label_zh: str | None = None
    description: str | None = None
    description_zh: str | None = None
    value_type: str
    required: bool

class TriggerEventsResponse(BaseModel):
    data: list[TriggerEvent]

class MessageVariant(BaseModel):
    title: str | None = None
    content: str | None = None
    content_type: str | None = None
    plain_content: str | None = None
    link: str | None = None
    cover: str | None = None


class Message(BaseModel):
    title: str
    content: str | None = None
    content_type: str | None = None
    plain_content: str | None = None
    link: str | None = None
    cover: str | None = None
    variants: dict[str, MessageVariant] | None = None

class NotificationSourceProvided(BaseModel):
    id: int
    uuid: str
    name: str
    name_zh: str
    category: str | None
    description: str | None
    description_zh: str | None
    create_time: datetime
    update_time: datetime | None

class NotificationTargetProvided(BaseModel):
    id: int
    uuid: str
    name: str
    name_zh: str
    category: str | None
    description: str | None
    description_zh: str | None
    create_time: datetime
    update_time: datetime | None

class NotificationSource(BaseModel):
    id: int
    title: str
    description: str | None
    notification_source_provided: NotificationSourceProvided
    create_time: datetime
    update_time: datetime | None
    creator: UserPublicInfo
    is_forked: bool | None = None
    is_public: bool

class NotificationSourceDetail(BaseModel):
    id: int
    title: str
    description: str | None
    notification_source_provided: NotificationSourceProvided
    create_time: datetime
    update_time: datetime | None
    config_json: str | None = None
    creator: UserPublicInfo
    is_public: bool

class NotificationTarget(BaseModel):
    id: int
    title: str
    notification_target_provided: NotificationTargetProvided
    description: str | None
    create_time: datetime
    update_time: datetime | None
    creator: UserPublicInfo
    is_forked: bool | None = None
    is_public: bool

class NotificationTargetDetail(BaseModel):
    id: int
    title: str
    notification_target_provided: NotificationTargetProvided
    description: str | None
    create_time: datetime
    update_time: datetime | None
    config_json: str | None = None
    creator: UserPublicInfo
    is_public: bool

class NotificationTargetsProvidedResponse(BaseModel):
    data: list[NotificationTargetProvided]

class NotificationTargetDetailRequest(BaseModel):
    notification_target_id: int

class NotificationSourceDetailRequest(BaseModel):
    notification_source_id: int

class NotificationTargetsResponse(BaseModel):
    data: list[NotificationTarget]

class NotificationSourcesProvidedResponse(BaseModel):
    data: list[NotificationSourceProvided]

class EmailTargetForm(BaseModel):
    email: str
    code: str

class IOSTargetForm(BaseModel):
    device_token: str

class FeiShuTargetForm(BaseModel):
    webhook_url: str
    sign: str

class DingTalkTargetForm(BaseModel):
    webhook_url: str
    sign: str

class TelegramTargetForm(BaseModel):
    chat_id: str

class AddNotificationTargetRequest(BaseModel):
    notification_target_provided_id: int
    title: str
    is_public: bool
    description: str | None = None
    email_target_form: EmailTargetForm | None = None
    ios_target_form: IOSTargetForm | None = None
    feishu_target_form: FeiShuTargetForm | None = None
    dingtalk_target_form: DingTalkTargetForm | None = None
    telegram_target_form: TelegramTargetForm | None = None

class EmailSourceForm(BaseModel):
    host: str
    port: int
    username: str
    password: str

class IOSSourceForm(BaseModel):
    team_id: str
    key_id: str
    private_key: str
    apns_topic: str

class FeiShuSourceForm(BaseModel):
    app_id: str
    app_secret: str

class TelegramSourceForm(BaseModel):
    bot_token: str

class AddNotificationSourceRequest(BaseModel):
    notification_source_provided_id: int
    title: str
    is_public: bool
    description: str | None = None
    email_source_form: EmailSourceForm | None = None
    ios_source_form: IOSSourceForm | None = None
    feishu_source_form: FeiShuSourceForm | None = None
    telegram_source_form: TelegramSourceForm | None = None

class UpdateNotificationSourceRequest(BaseModel):
    notification_source_id: int
    title: str | None = None
    description: str | None = None
    is_public: bool | None = None
    email_source_form: EmailSourceForm | None = None
    ios_source_form: IOSSourceForm | None = None
    feishu_source_form: FeiShuSourceForm | None = None
    telegram_source_form: TelegramSourceForm | None = None

class UpdateNotificationTargetRequest(BaseModel):
    notification_target_id: int
    title: str | None = None
    description: str | None = None
    is_public: bool | None = None
    email_target_form: EmailTargetForm | None = None
    ios_target_form: IOSTargetForm | None = None
    feishu_target_form: FeiShuTargetForm | None = None
    dingtalk_target_form: DingTalkTargetForm | None = None
    telegram_target_form: TelegramTargetForm | None = None

class DeleteNotificationTargetRequest(BaseModel):
    notification_target_ids: list[int]

class DeleteNotificationSourceRequest(BaseModel):
    notification_source_ids: list[int]

class NotificationRecordDetailRequest(BaseModel):
    notification_record_id: int

class SearchNotificationRecordRequest(BaseModel):
    keyword: str | None = None
    start: int | None = None
    limit: int = Field(default=10, le=PUBLIC_PAGINATION_LIMIT)

class ReadNotificationRecordRequest(BaseModel):
    notification_record_ids: list[int]
    status: bool

class NotificationRecord(BaseModel):
    id: int
    title: str
    content: str | None
    read_at: datetime | None
    link: str | None
    cover: str | None
    creator: UserPublicInfo
    create_time: datetime
    update_time: datetime | None

class DeleteNotificationRecordRequest(BaseModel):
    notification_record_ids: list[int]

class NotificationTaskDetailRequest(BaseModel):
    notification_task_id: int

class NotificationTriggerEvent(BaseModel):
    trigger_event_id: int

class NotificationTriggerScheduler(BaseModel):
    cron_expr: str

class NotificationTask(BaseModel):
    id: int
    creator_id: int
    title: str
    enable: bool
    content_type: int
    trigger_type: int
    trigger_event: NotificationTriggerEvent | None = None
    trigger_scheduler: NotificationTriggerScheduler | None = None
    notification_title: str | None = None
    notification_content: str | None = None
    notification_link: str | None = None
    notification_cover: str | None = None
    notification_template_id: int | None = None
    notification_template_bindings: dict[str, 'NotificationTemplateBinding'] | None = None
    notification_source: NotificationSource | None = None
    notification_target: NotificationTarget | None = None
    create_time: datetime
    update_time: datetime | None

class DeleteNotificationTaskRequest(BaseModel):
    notification_task_ids: list[int]

class UpdateNotificationTaskRequest(BaseModel):
    notification_task_id: int
    title: str | None = None
    content_type: int | None = None
    enable: bool | None = None
    notification_template_id: int | None = None
    notification_template_bindings: dict[str, 'NotificationTemplateBinding'] | None = None
    trigger_type: int | None = None
    trigger_scheduler_cron: str | None = None
    trigger_event_id: int | None = None
    notification_title: str | None = None
    notification_content: str | None = None
    notification_link: str | None = None
    notification_cover: str | None = None
    notification_source_id: int | None = None
    notification_target_id: int | None = None

class AddNotificationTaskRequest(BaseModel):
    notification_source_id: int
    notification_target_id: int
    enable: bool
    title: str
    content_type: int
    notification_template_id: int | None = None
    notification_template_bindings: dict[str, 'NotificationTemplateBinding'] | None = None
    notification_title: str | None = None
    notification_content: str | None = None
    notification_cover: str | None = None
    notification_link: str | None = None
    trigger_type: int
    trigger_event_id: int | None = None
    trigger_scheduler_cron: str | None = None

class NotificationTemplate(BaseModel):
    id: int
    uuid: str
    creator_id: int
    is_public: bool = False
    name: str
    description: str | None = None
    title_template: str | None = None
    content_template: str | None = None
    link_template: str | None = None
    cover_template: str | None = None
    create_time: datetime
    update_time: datetime | None = None
    creator: UserPublicInfo
    is_forked: bool | None = None
    parameters: list['NotificationTemplateParameter'] = []

class NotificationTemplateParameter(BaseModel):
    id: int
    key: str
    label: str
    description: str | None = None
    value_type: str
    required: bool
    default_value: str | None = None

class NotificationTemplateBinding(BaseModel):
    source_type: Literal['event', 'static']
    attribute_key: str | None = None
    static_value: str | None = None

class NotificationTemplateUpsertRequest(BaseModel):
    notification_template_id: int | None = None
    name: str
    is_public: bool = False
    description: str | None = None
    title_template: str
    content_template: str | None = None
    link_template: str | None = None
    cover_template: str | None = None
    parameters: list['NotificationTemplateParameterUpsertRequest'] = []

class NotificationTemplateParameterUpsertRequest(BaseModel):
    key: str
    label: str
    description: str | None = None
    value_type: str = 'string'
    required: bool = False
    default_value: str | None = None

class DeleteNotificationTemplateRequest(BaseModel):
    notification_template_id: int

class NotificationTemplatesResponse(BaseModel):
    data: list[NotificationTemplate]

class NotificationSourcesUsableResponse(BaseModel):
    data: list[NotificationSource]

class NotificationTargetsUsableResponse(BaseModel):
    data: list[NotificationTarget]
